### DISCUSSION TITLE
The explaination is wrong in typescript guide

### DISCUSSION THREAD

**Discussion by GuichiZhao** [pmndrs/zustand#2895] [Bug report] [answered]
I believe the section [Basic usage](https://zustand.docs.pmnd.rs/guides/typescript#basic-usage) from the typescript guide is completely wrong

The reason why "Why can't we simply infer the type from the initial state" is wrong per the typescript guide, it has **nothing to do with invariant**, the real reason is that we **do not provide any useful information** for TS engine does the pattern matching

If we provide proper information
```
declare const create: <T>(f: (get: () => T) => T) => T
const a = create(() => 'hello') //const a: string
```
We can see clearly that a is inferred as a `number` rather than `unknown` as the typescript guide said. 

> So, as long as the generic to be inferred is invariant (i.e. both covariant and contravariant), TypeScript will be unable to infer it.

 Let me speak aloud about how TS engine think

When the typescript engine sees `() => 'hello'` it tries to pattern match `(get: () => T) => T` and it figures out that if it binds `T` with `string`, it will match both.
Note that because the `()=>'hello'` provides no param, the TS engine will not bother to match the argument, it will just let it go
Remember that `()=>sting` is perfectly assignable to `(a:number)=>string` (but no the other way)

However, if we put an argument as a get param,
```
declare const create: <T>(f: (get: () => T) => T) => T
const a = create((_) => 'hello') // const a: unknown
```

TS engine can not bind T properly, it is not because of a chicken and egg problem as the guide said

> TypeScript wonders. It's like that chicken or egg problem. At the end TypeScript, gives up and infers T as unknown.

Let us think aloud again, the TS engine tries to match
`(_) => 'hello'` and `(get: () => T) => T` it sees the parameter is not empty as in the previous example, so it **DOES** try to match it, so the problem becomes:
match `_` and `() => T`, where `_` is implicitly of `any` type.  there is no exra hint with `_` at all, so `<T>` is inferred as `unknown`

If we annotate it to hint the typescript, it works again
```
declare const create: <T>(f: (get: () => T) => T) => T
const a = create((_: (s: string) => string) => 'hello') //const a: unknown
```

Similar behavior can be observed in Promise
```
const x = new Promise((resolve) => {
  resolve(11)
}) //const x: Promise<unknown>
```
The Promise can not be bound properly because when TS tries to pattern match the arguments, it finds no hint at all, **AGAIN nothing to do with "invariant"**

If we provide the TS engine with some hint, it will work again
```
const x = new Promise((resolve: (s: number) => void) => {
  resolve(11)
}) //const x: Promise<number>
```

**To conclude, the real reason why TS engine infers `<T>` as `unknown` is there is no enough information provided, rather the the "chicken-egg" or "invariant" as the documentation put it.** 

Reproduction Link
https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEYD2A7AzgF0XKGQC54AeAFQD4AKAM0IoHMQNaBKeAXjPhNY6585IAoZOixR22ELhAUKAfVppCmGAEsUdfvBXrN7TgHIAFiAgQkB5oOsjM8AB4SUIAO7wACjCQBbVWhkUcGhIEABuBPAUSvAoAK7eAEYgMFqhSKrAWgDegvDwQSHhFACMxVYAvsxAA

If the idea is approved, I am happy to issue a PR to fix the documentation

**Comment by dai-shi** [comment]
Are you suggesting to just fix the docs, or do you have any idea to improve types in the code?

**Reply by GuichiZhao** [reply]
> Are you suggesting just fixing the docs, or do you have any idea to improve types in the code?

I have no idea about how to improve the types in the code. But I am glad to rewrite the documentation,

As far as I know, all the TS engine cares about is **pattern matching**. As long as the pattern can be matched, the engine will simply bind the `<T>` to a specific type. It **doesn't matter** where the `<T>` place, is it can be a parameter, or a return value, or a return value of a parameter,  or the parameter of the parameter,  or a ...  of a .. of a ... etc

So, it will be very misleading to think how the `<T>` is used, it can be circularly referenced, but it will only affect runtime behavior as the current documentation pointed out, without messing up compile time behavior, i.e. TS engine.

The documentation should highlight that if we want to TS engine to help us refer to the type properly, we need to provide it with enough hints, rather than approaching the problem from how the runtime behaves (circular reference, or chicken/egg problem), runtime and compile time are different logics.

If you agree with what I said, I can draft a new version of the documentation of the "Basic usage" section.

@dai-shi @bjornstar

**Reply by dai-shi** [reply]
I think you want to talk with @devanshj in advance, as he's the one who created the TypeScript guide.

**Reply by GuichiZhao** [reply]
> I think you want to talk with @devanshj in advance, as he's the one who created the TypeScript guide.

What is your opinion though?

**Reply by dai-shi** [reply]
Apart from the wording, I think it feels like you are missing some points of code and/or docs. I can't explain it very well. Maybe #725 will help. And, #710. I don't understand them 100% and that's why it's better for you guys to discuss.

**Reply by GuichiZhao** [reply]
> Apart from the wording, I think it feels like you are missing some points of code and/or docs. I can't explain it very well. Maybe #725 will help. And, #710. I don't understand them 100% and that's why it's better for you guys to discuss.

If you can not fully understand, How can expect the user to understand the documentation? LOL


By the way, I am 90% sure the real reason you do not understand it is that the current documentation is not only wrong but also confusing and misleading. It points in the wrong direction. That is why no one can follow its logic

If you read my text above, you will find it easier to follow and make more sense

**Reply by dai-shi** [reply]
I mean about #710 and #725. If it's about the wording in the guide, I think I can understand. Let's see if @devanshj catches up, otherwise will see.

**Comment by devanshj** [comment, marked as answer]
Hi, I'm the author of the said documentation. Thanks for pointing out it's not because of invariance but because of the naive pattern matching TypeScript uses (as illustrated in the promise example too). Please go ahead and make a PR to improve the documentation.

**Reply by GuichiZhao** [reply]
I tried to contribute to the documentation and follow the [contribution guide](https://github.com/pmndrs/zustand/blob/main/CONTRIBUTING.md). , but I faced some trouble setting up the environment

<img width="1586" alt="Screenshot 2024-12-14 at 15 46 53" src="https://github.com/user-attachments/assets/07902305-86a4-4eb4-8cad-64af89d78dea" />

1. There is no branch names `docs` in the website repo

<img width="496" alt="Screenshot 2024-12-14 at 15 51 46" src="https://github.com/user-attachments/assets/537459e8-fc6f-4e41-8ad0-539c9599dec6" />

2. There is no `src` folder as well
<img width="506" alt="Screenshot 2024-12-14 at 15 51 58" src="https://github.com/user-attachments/assets/ad5f7b96-13b2-4bdd-9efb-29c27983fcc8" />

@devanshj

**Reply by dai-shi** [reply]
I'm VERY sorry about it. CONTRIBUTING.md isn't well maintained. ( @dbritto-dev Can you fix it quickly? )

@GuichiZhao It's here: https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md
Just open a PR in this repo.

**Reply by GuichiZhao** [reply]
@dai-shi  Here is the PR https://github.com/pmndrs/zustand/pull/2900

**Reply by dbritto-dev** [reply]
Hey peeps I'll update `CONTRIBUTING.md`

**Comment by dai-shi** [comment]
@devanshj Can you please review #2900?

**Comment by heecheon92** [comment]
I think the sample code itself is the real cause of the confusion for the typescript compiler.

```ts
declare const create: <T>(f: (get: () => T) => T) => T

const x = create((get) => ({
  foo: 0,
  bar: () => get(),
}))
```

writing above is like the following at the type level,

```ts
create((get) => ({
    foo: 0,
    bar: () => {
      foo: 0,
      bar: () => {
        foo: 0,
        bar: () => { ... so on},
      },
    },
}))
```

- The function create is generic <T>(f: (get: () => T) => T) => T.
- Inside f, we return an object of type { foo: number; bar: () => T }.
- But that same T is also the return type of get().

So “what is T?” TypeScript sees that:

1. f returns an object that must be typed as T.
2. But f also calls get(), whose return type is T.
3. Hence “T is something that has bar: () => T, but that same T is also returned by get() itself”—a cyclical definition.

So in a sense, it, indeed, is a "chicken of egg" problem, not "chicken or egg" from variance perspective, but cyclical reference perspective.

@GuichiZhao is right for the simpler case where a parameter is completely untyped (like (param) => "hello") and TS truly has no info to unify the generic.

However, cyclical references also break TS’s inference, even if you are “returning an object shape that references T.” That shape itself can be insufficient because TS sees a self-referential definition, not just “lack of hints.”
